WordPress/src/main/java/org/wordpress/android/ui/uploads/AutoSavePostIfNotDraftUseCase.kt

```diff
@@ -0,0 +1,144 @@
+        remotePostPayload: RemotePostPayload,
+        callback: OnAutoSavePostIfNotDraftCallback
+    ) {
+        val localPostId = LocalId(remotePostPayload.post.id)
+        if (remotePostPayload.post.isLocalDraft) {
+            throw IllegalArgumentException("Local drafts should not be auto-saved")
+        }
+        if (postStatusContinuations.containsKey(localPostId) ||
+                autoSaveContinuations.containsKey(localPostId)) {
+            throw IllegalArgumentException(
+                    "This post is already being processed. Make sure not to start an autoSave " +
+                            "or update draft action while another one is going on."
+            )
+        }
+        GlobalScope.launch(bgDispatcher) {
+            val onPostStatusFetched = fetchRemotePostStatus(remotePostPayload)
+            val result = when {
+                onPostStatusFetched.isError -> {
+                    FetchPostStatusFailed(
+                            post = remotePostPayload.post,
+                            error = onPostStatusFetched.error
+                    )
+                }
+                onPostStatusFetched.remotePostStatus == DRAFT_POST_STATUS -> {
+                    PostIsDraftInRemote(remotePostPayload.post)
+                }
+                else -> {
+                    autoSavePost(remotePostPayload)
+                }
+            }
+            callback.handleAutoSavePostIfNotDraftResult(result)
+        }
+    }
+
+    private suspend fun fetchRemotePostStatus(remotePostPayload: RemotePostPayload): OnPostStatusFetched {
+        val localPostId = LocalId(remotePostPayload.post.id)
+        return suspendCancellableCoroutine { cont ->
+            postStatusContinuations[localPostId] = cont
+            dispatcher.dispatch(PostActionBuilder.newFetchPostStatusAction(remotePostPayload))
+        }
+    }
+
+    private suspend fun autoSavePost(remotePostPayload: RemotePostPayload): AutoSavePostIfNotDraftResult {
+        val localPostId = LocalId(remotePostPayload.post.id)
+        val onPostChanged: OnPostChanged = suspendCancellableCoroutine { cont ->
+            autoSaveContinuations[localPostId] = cont
+            dispatcher.dispatch(PostActionBuilder.newRemoteAutoSavePostAction(remotePostPayload))
+        }
+        return if (onPostChanged.isError) {
+            PostAutoSaveFailed(remotePostPayload.post, onPostChanged.error)
+        } else {
+            val updatedPost = postStore.getPostByLocalPostId(localPostId.value)
+            PostAutoSaved(updatedPost)
+        }
+    }
+
+    @Subscribe(threadMode = BACKGROUND)
+    @Suppress("unused")
+    fun onPostStatusFetched(event: OnPostStatusFetched) {
+        val localPostId = LocalId(event.post.id)
+        postStatusContinuations[localPostId]?.let { continuation ->
+            continuation.resume(event)
+            postStatusContinuations.remove(localPostId)
+        }
+    }
+
+    @Subscribe(threadMode = BACKGROUND)
```

Comment:
I have made this change in bb985b123f331f84691c7101cc118ae5bc968bdf, but I couldn't test it because I can't reproduce the issue even before the change. Also, as discussed on Slack, I think there is a better way to handle the priority issue, so I'll open an issue for that once this PR is merged.